process/process.go

```diff
@@ -749,7 +732,12 @@ func (p *Process) insertPropose(propose Propose) bool {
 		return false
 	}
 
+	if _, ok := p.ProposeLogs[propose.Round]; ok {
+		return false
+	}
+
 	p.ProposeLogs[propose.Round] = propose
+	p.TraceLogs[propose.Round][propose.From] = true
```

Comment:
As per the previous suggestion, I recommend moving `TraceLogs` initialisation to here.
```suggestion
	if _, ok := p.ProposeLogs[propose.Round]; ok {
		return false
	}
	p.ProposeLogs[propose.Round] = propose

	if _, ok := p.TraceLogs[propose.Round]; !ok {
		p.TraceLogs[propose.Round] = map[id.Signatory]bool{}
	}
	p.TraceLogs[propose.Round][propose.From] = true
```